doc/explanation/package-management.md

```diff
@@ -0,0 +1,220 @@
+with the current one. Thus given the same inputs it should always have the same
+outputs; no state is held between the solver runs.
+
+This can lead to more packages being installed (as OPAM won't by default
+install new versions of packages if the existing versions satisfy the
+constraints) but it avoids interference from already installed packages that
+lead to potentially different solutions.
+
+After solving is done, the solution gets written into the lock directory with
+all the data that is necessary to build and install the packages. From this
+point, the project contains all necessary information and does not need to
+access any package repositories.
+
+## Building
+
+When building, Dune will read the information from the lock directory and set
+up build rules to use the packages. The rules include
+
+* Fetch rules to download the sources as well as any additional sources like
+  patches and unpack them
+* Build rules to evaluate the build instructions from the build instructions
+  stored in the lock directory
+* Install rules to put the artifacts that were built into the appropriate,
+  Dune-managed folders
+
+Creating these processes as rules means that they will only be executed on
+demand, thus if the project has already downloaded the sources, it does not
+need to download them again. Likewise, if packages are installed, they stay
+installed.
+
+The results of the rules are stored in the `_build` directory of the project
+and managed automatically by Dune. Thus when cleaning the build directory, the
+installed packages are cleaned as well and will be reinstalled at the next
+build.
+
+When building the users project, the installed packages are added to the
+necessary search paths, thus user code can use the dependencies without any
+additional configuration.
+
+(compatibility)=
+## Packaging for Dune Compatibility
+
+Dune can build and install most packages as dependencies, even if they are not
+built with Dune themselves. Dune will execute the build instructions from the
+lock file, very similar to `opam`.
+
+However, there are some important restrictions that packages must adhere to to
+be able to work with Dune.
+
+The most important one is that the packages must not use absolute paths to
+refer to files. That means, they cannot read the path they are being built or
+installed into and expect this path to be stable. Dune builds packages in a
+sandbox location and after the build has finished it moves the files to the
+actual destination.
+
+The reason for this is that one hand it enables building without messing up the
+current state and on the other hand it allows for caching artifacts across
+projects.
+
+To sidestep these restructions in many cases the solution is to use relative
+paths, as Dune guarantees that packages installed into different sections are
+installed in a way where their relative location stays the same.
+
+A minor difference is that Dune does not support packages installing themselves
```

Comment:
I'm not sure to understand this paragraph, could you elaborate?

Reply:
Some (very few) opam packages install themselves in the compilers `stdlib` folder, where you don't need to declare them as dependency in `(libraries)` (so they work like `List` or `String` etc).

This doesn't work in Dune, as Dune does not allow packages to install themselves into other package's folders. It is a bit hacky and only works in OPAM because the compiler's stdlib folder has a known location; OPAM does not expose that location as a section for `install` files.

Do you have a suggestion how to describe this in an understandable way?

Reply:
Out of curiosity, do you have a package name doing this? :smile: 

Otherwise, the explanation you gave me is pretty understandable. I was lacking the context that some packages are actually doing this! I think providing the context would make it more understandable. 
```
Some packages install themselves in the Stdlib directory
-> Allow by opam because of a known location 
-> Dune is per project so don't have a global location
-> [rest of you explanation]
```
:+1:

Reply:
Yes, we encountered this issue with the `num` package: https://github.com/ocaml/opam-repository/pull/26178. I think the only packages that would do this are packages that have been spun out of the stdlib (thus were previously accessible without a `libraries` entry thus the `compat` packages preserve the behavior) but don't install a into a package directory. In the case of `num` we solved this by installing into both places and that seems to be a best-of-both-worlds situation.

However, this is a fairly rare case, as `num` was the only occurrence, so maybe not even worth mentioning.